proxy/src/main.rs

```diff
@@ -0,0 +1,163 @@
+use anyhow::{anyhow, bail, Result};
+use reqwest::blocking::{Client, Response};
+use reqwest::header::{self, HeaderMap};
+use reqwest::Method;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::process::ExitCode;
+use std::str::FromStr;
+use std::time::Duration;
+use std::{env, io};
+use url::Url;
+
+const ENV_CONFIG: &str = "SD_PROXY_ORIGIN";
+
+/// Incoming requests (as JSON) received over stdin
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct IncomingRequest {
+    method: String,
+    path_query: String,
+    stream: bool,
+    #[serde(default)]
+    headers: HashMap<String, String>,
+    body: Option<String>,
+    #[serde(default = "default_timeout")]
+    timeout: u64,
+}
+
+/// Default timeout for requests; serde requires this be a function
+fn default_timeout() -> u64 {
+    10
+}
+
+/// Serialization format for non-streamed requests
+#[derive(Serialize, Debug)]
+struct OutgoingResponse {
+    status: u16,
+    headers: HashMap<String, String>,
+    body: String,
+}
+
+/// In a streamed response, we emit the checksum after the streaming finishes
+#[derive(Serialize, Debug)]
+struct StreamMetadataResponse {
+    sha256sum: String,
+    filename: String,
+}
+
+/// Serialization format for errors, always over stderr
+#[derive(Serialize, Debug)]
+struct ErrorResponse {
+    error: String,
+}
+
+fn handle_json_response(resp: Response) -> Result<()> {
+    let mut headers = HashMap::new();
+    for (name, value) in resp.headers() {
+        headers.insert(name.to_string(), value.to_str()?.to_string());
+    }
+    let outgoing_response = OutgoingResponse {
+        status: resp.status().as_u16(),
+        headers,
+        body: resp.text()?,
+    };
+    println!("{}", serde_json::to_string(&outgoing_response)?);
+    Ok(())
+}
+
+fn handle_stream_response(mut resp: Response) -> Result<()> {
+    // Get the headers, will be output later but we want to fail early if it's missing/invalid
+    let etag = resp
+        .headers()
+        .get(header::ETAG)
+        .ok_or_else(|| anyhow!("missing etag header"))?
+        .to_str()?
+        .to_string();
+    let content_disposition = resp
+        .headers()
+        .get(header::CONTENT_DISPOSITION)
+        .ok_or_else(|| anyhow!("missing content-disposition"))?
+        .to_str()?
+        .to_string();
+    let mut stdout = io::stdout().lock();
+    resp.copy_to(&mut stdout)?;
```

Comment:
I missed this on the first ten read-through and was confused, so I hereby suggest making this explicit too:  :-)

```suggestion
    // Emit the streamed response as stdout
    let mut stdout = io::stdout().lock();
    resp.copy_to(&mut stdout)?;
```

Reply:
This has been replaced, but in the new code I've tried to make it clear where the streaming is happening